state/pruner.go

```diff
@@ -0,0 +1,343 @@
+	}
+	currentRetainHeight, err := p.stateStore.GetABCIResRetainHeight()
+	if err != nil {
+		if !errors.Is(err, ErrKeyNotFound) {
+			return err
+		}
+		return p.stateStore.SaveABCIResRetainHeight(height)
+	}
+	if currentRetainHeight > height {
+		return ErrPrunerCannotLowerRetainHeight
+	}
+	return p.stateStore.SaveABCIResRetainHeight(height)
+}
+
+// GetApplicationRetainHeight is a convenience method for accessing the
+// GetApplicationRetainHeight method of the underlying state store.
+func (p *Pruner) GetApplicationRetainHeight() (int64, error) {
+	return p.stateStore.GetApplicationRetainHeight()
+}
+
+// GetCompanionBlockRetainHeight is a convenience method for accessing the
+// GetCompanionBlockRetainHeight method of the underlying state store.
+func (p *Pruner) GetCompanionBlockRetainHeight() (int64, error) {
+	return p.stateStore.GetCompanionBlockRetainHeight()
+}
+
+// GetABCIResRetainHeight is a convenience method for accessing the
+// GetABCIResRetainHeight method of the underlying state store.
+func (p *Pruner) GetABCIResRetainHeight() (int64, error) {
+	return p.stateStore.GetABCIResRetainHeight()
+}
+
+func (p *Pruner) pruneABCIResRoutine() {
+	p.logger.Info("Started pruning ABCI results", "interval", p.interval.String())
+	lastABCIResRetainHeight := int64(0)
+	for {
+		select {
+		case <-p.Quit():
+			return
+		default:
+			newABCIResRetainHeight := p.pruneABCIResToRetainHeight(lastABCIResRetainHeight)
+			p.observer.PrunerPrunedABCIRes(&ABCIResponsesPrunedInfo{
+				FromHeight: lastABCIResRetainHeight,
+				ToHeight:   newABCIResRetainHeight - 1,
+			})
+			lastABCIResRetainHeight = newABCIResRetainHeight
+			time.Sleep(p.interval)
+		}
+	}
+}
+func (p *Pruner) pruneBlocksRoutine() {
+	p.logger.Info("Started pruning blocks", "interval", p.interval.String())
+	lastRetainHeight := int64(0)
+	for {
+		select {
+		case <-p.Quit():
+			return
+		default:
+			newRetainHeight := p.pruneBlocksToRetainHeight(lastRetainHeight)
+			p.observer.PrunerPrunedBlocks(&BlocksPrunedInfo{
+				FromHeight: lastRetainHeight,
+				ToHeight:   newRetainHeight - 1,
+			})
+			lastRetainHeight = newRetainHeight
+			time.Sleep(p.interval)
+		}
+	}
+}
+
+func (p *Pruner) pruneBlocksToRetainHeight(lastRetainHeight int64) int64 {
+	targetRetainHeight := p.findMinRetainHeight()
+	if targetRetainHeight == lastRetainHeight {
+		return lastRetainHeight
+	}
+	pruned, evRetainHeight, err := p.pruneBlocks(targetRetainHeight)
+	// The new retain height is the current lowest point of the block store
+	// indicated by Base()
+	newRetainHeight := p.bs.Base()
+	if err != nil {
+		p.logger.Error("Failed to prune blocks", "err", err, "targetRetainHeight", targetRetainHeight, "newRetainHeight", newRetainHeight)
+	} else if pruned > 0 {
+		p.logger.Info("Pruned blocks", "count", pruned, "evidenceRetainHeight", evRetainHeight, "newRetainHeight", newRetainHeight)
+	}
+	return newRetainHeight
+}
+
+func (p *Pruner) pruneABCIResToRetainHeight(lastRetainHeight int64) int64 {
```

Comment:
It should be ok even if it was changed between then.. because changing it can only happen if the new height is higher than the previous one, which means we will still prune the previously requested results. Same for block results. 

For example the retain height is set to 5. 
We enter `pruneABCIResToRetaineHeight` at the same time someone tries to write 3. That someone won't be allowed to write 3 due to the check in 186.  If they try to set it to 6, and we read it just after that,it's still ok.